ddev/ddev
# feat: adds support to ephemeral ports for 80/443, fixes #5326

## The Issue

- #5326 

Ports 80 and 443 and commonly occupied. DDEV should be able to successfully start by creating ephemeral ports as an alternative.

## How This PR Solves The Issue
New functionality is added that checks as soon as possible if the ports HTTP and HTTPS, used by the router, are occupied.
If they are occupied, a ephemeral port is assigned and used for configuring the router.

The router registers if it's using ephemeral ports in an environment variable. This information is readed by other projects that start and is useful to detect if the router itself is occuping the http and https ports, or not. 

# TODO

- [x] `ddev describe` does not show to correct URLs in the bottom line "All URLs"
- [x] `router_http_port` and `router_https_port` are not updated in the project. Should they be? Or should this just be temporary? We do see the expected:
    > HTTP port 28080 is busy, using 50080 instead.
    > HTTPS port 28443 is busy, using 50443 instead.
- [ ] `app.Stop()` pushes [removes site config from router](https://github.com/ddev/ddev/blob/c91869a5a840f55eeb020d71d09e80158542dcfc/pkg/ddevapp/ddevapp.go#L2568-L2575). This is good, and may be enough. However, the router itself still is configured with an entrypoint (no longer used, perhaps) in ~/.ddev/traefik/.static_config.yaml. And in this case the router is still running bound to the ports that are perhaps no longer in use. To demonstrate: `docker exec -it ddev-router cat .static_config.yaml`. This is not a bug related to the work in this PR, may not be a bug at all but it may be related to some of the weirdness we see in other tests after the ephemeral tests run.
- [x] The logic has to handle *all* ports handled by the router, not just http/https. 
- [x] There's no real reason to distinguish between HTTP and HTTPS I don't think. We just need the ports.
- [x] Handle the mailpit ports at least
- [x] Handle HTTP_EXPOSE and HTTPS_EXPOSE related to `web_extra_exposed_ports`, a very common scenario
- [x] Consider if there's a way to handle conflicts on 3rd party services, which would require overriding their HTTP_EXPOSE and HTTPS_EXPOSE.
- [x] I removed the second site start from TestUseEphemeralPort, but it's a good thing to have that second site, as I've discovered while breaking things. It should go back in there.
- [x] Don't use "80", "443", "8025" and "8026" as literals. Make those global and configurable. We already have constants for these in values.go, DdevDefaultRouterHTTPPort and DdevDefaultRouterHTTPSPort etc. But we can make them variables, and override them with a DDEV_TEST environment variable, so that multiple non-conflicting tests might possibly be runnable at the same time. (followup in https://github.com/ddev/ddev/pull/6470)
- [x] Comment out or make DDEV_VERBOSE all the massive output showing HTTP_EXPOSE etc.


## Manual Testing Instructions
First, make sure any DDEV container is off by running:
`ddev poweroff`

Then, occupy the 80 and 443 ports. For example, by running:
`docker run -d -p 80:80 -p 443:443 nginx`

Then, go to a DDEV project that is configured to use the same ports. This is managed in the config variables:
router_http_port="80"
router_https_port="443"

Then run
`ddev start`
You will see a message, very early on, indicating that the ports are occupied and an ephemeral port wil be used instead.
The start process should finish normally, and DDEV already indicates where to find the project by displaying the URL.
You can also run
`ddev describe`
to see information about the URL.

Then, repeat the same operation with a different project that is also configured to use the same ports.
DDEV should detect the ephemeral ports and assign them to the new project.



## Automated Testing Overview

There are 2 tests created:
- One for checking that DDEV is able to find an available port in a given range.
- Another for checking that the app get's the ephemeral port when it has to, and that a second app will also get the same ports.

## Release/Deployment Notes

This feature should not have any negative effect in other parts of the code.
